File: code/R/src/rl_agent/tdva.R
# An Reinforcement Learning Application for Statistical Arbitrage
# karol.przybylak.13@ucl.ac.uk, kmprzybylak@gmail.com

# Qlearning

#moze agent moze uzywac t-1 i t, zamiast t i t+1



# -----------------------------------------------------------------
#region Dependencies

#endregion
# -----------------------------------------------------------------
#region Function Definition

tdva_control <- function(pair, coef, features, Nepisodes, cost, pretrained_agent=NULL, ...){
  # Initialize all elements of the RL framework
  params <- list(...)

  list[actions, f_params, alpha, epsilon, N_0, episode, Action_track] <- InitializeRLframework(features, 'tdva')

  # If available, continue with the policy from the training period
  if(!is.null(pretrained_agent)) list[actions, f_params, alpha, epsilon, N_0, episode, Action_track] <- pretrained_agent

  while(episode <= Nepisodes)
  {
   i <- 1
    while (i < nrow(pair))
    {

    state <- features[i,]
    action <- ChooseActionTD(f_params, state, epsilon, actions) # choose action accordin to the present state of FA (separate model for each action)
    action_dim <- match(action, actions)
    Qsa <- c(f_params %*% t(as.matrix(state)))[action_dim]
    state_action <- as.numeric(c(state, action_dim))

    Action_track <- c(Action_track, action)
    Action_count <- table(Action_track)
    a_count = Action_count[names(Action_count) == action]

    # Update constants
    # EPsilon should also depend on the value of the present state, could grow bigger if negative state values
    # But maybe they have been evaluated to such a value? But maybe it should grow with growing volatility or if
    # a regime switch was detected. Or otherwise: sudden changes in values indicate changing conditions and should
    # create a bigger epsilon again.
    epsilon <- N_0 / (N_0 + i)
    #alpha <- 1/Nsa[matrix(state_action, 1)]
    alpha <- N_0 / (N_0 + a_count)

    state_next <- features[i+1,] # get the following state
    action_target <- ChooseActionTD(f_params, state_next, epsilon, actions) # choose the next action with the same policy (actualy should also use old epsilon)
    action_target_dim <- match(action_target, actions)
    Qsa_target <- c(f_params %*% t(as.matrix(state)))[action_target_dim]
    state_action_target <- as.numeric(c(state_next, action_target_dim))

    # 1-step return
    ret <- Qreward(pair[c(i:(i+1)),], coef, action, cost)

    # Update the model
    delta = unlist(alpha*(ret + Qsa_target - Qsa)*state)
    f_params[action_dim,] <- f_params[action_dim,] + delta


    i <- i + 1
    }
   episode <- episode + 1
  }

  return(list(Qsa, Action_track, Nsa, Ns, epsilon, alpha, cut_points))
}

File: code/python/src/rl_agent/tdva.py
"""
Temporal Difference Value Approximation (TDVA) implementation.

This module implements the TDVA algorithm for reinforcement learning
in statistical arbitrage applications.
"""

from typing import Dict, Optional

import numpy as np
import pandas as pd
from src.rl_agent_python.reward import q_reward
from src.rl_agent_python.rl_utils import choose_action_td, initialize_rl_framework


def tdva_control(
    pair: pd.DataFrame,
    coef: np.ndarray,
    features: pd.DataFrame,
    n_episodes: int,
    cost: float,
    pretrained_agent: Optional[Dict] = None,
    **kwargs,
) -> Dict:
    """
    Control function for TDVA algorithm.

    Args:
        pair: DataFrame containing pair trading data
        coef: Array of cointegration coefficients
        features: DataFrame of state features
        n_episodes: Number of episodes to run
        cost: Transaction cost
        pretrained_agent: Optional dictionary containing pretrained agent parameters
        **kwargs: Additional parameters

    Returns:
        Dictionary containing:
        - Qsa: Q-values for state-action pairs
        - action_track: History of actions taken
        - nsa: Count of state-action pairs
        - ns: Count of states
        - epsilon: Final exploration rate
        - alpha: Final learning rate
        - cut_points: Discretization cut points
    """
    # Initialize RL framework
    (
        actions,
        f_params,
        alpha,
        epsilon,
        n_0,
        episode,
        action_track,
    ) = initialize_rl_framework(features, "tdva")

    # If available, continue with pretrained policy
    if pretrained_agent is not None:
        actions = pretrained_agent["actions"]
        f_params = pretrained_agent["f_params"]
        alpha = pretrained_agent["alpha"]
        epsilon = pretrained_agent["epsilon"]
        n_0 = pretrained_agent["n_0"]
        episode = pretrained_agent["episode"]
        action_track = pretrained_agent["action_track"]

    while episode <= n_episodes:
        i = 0
        while i < len(pair) - 1:
            # Get current state and choose action
            state = features.iloc[i]
            action = choose_action_td(f_params, state, epsilon, actions)
            action_dim = actions.index(action)

            # Calculate Q-value for current state-action
            qsa = np.dot(f_params[action_dim], state)
            state_action = list(state) + [action_dim]

            # Track actions and update counts
            action_track.append(action)
            action_count = pd.Series(action_track).value_counts()
            a_count = action_count.get(action, 0)

            # Update exploration and learning rates
            epsilon = n_0 / (n_0 + i)
            alpha = n_0 / (n_0 + a_count)

            # Get next state and target action
            state_next = features.iloc[i + 1]
            action_target = choose_action_td(f_params, state_next, epsilon, actions)
            action_target_dim = actions.index(action_target)
            qsa_target = np.dot(f_params[action_target_dim], state_next)
            state_action_target = list(state_next) + [action_target_dim]

            # Calculate 1-step return
            ret = q_reward(pair.iloc[i : i + 2], coef, action, cost)

            # Update the model
            delta = alpha * (ret + qsa_target - qsa) * state
            f_params[action_dim] += delta

            i += 1

        episode += 1

    return {
        "qsa": qsa,
        "action_track": action_track,
        "nsa": None,  # These would need to be tracked separately
        "ns": None,
        "epsilon": epsilon,
        "alpha": alpha,
        "cut_points": None,  # This would need to be passed in or calculated
    }
